lib.php

```diff
@@ -0,0 +1,286 @@
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Step subplugin to notify students of a course that the course is being deleted.
+ *
+ * @package    lifecyclestep_notifystudents
+ * @copyright  2021 Aaron Koßler WWU
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+namespace tool_lifecycle\step;
+
+use context_course;
+use core_user;
+use tool_lifecycle\local\manager\process_manager;
+use tool_lifecycle\local\manager\settings_manager;
+use tool_lifecycle\local\manager\step_manager;
+use tool_lifecycle\local\response\step_response;
+use tool_lifecycle\settings_type;
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once(__DIR__ . '/../lib.php');
+
+/**
+ * Step subplugin to notify students of a course that the course is being deleted.
+ *
+ * @package    lifecyclestep_notifystudents
+ * @copyright  2021 Aaron Koßler WWU
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class notifystudents extends libbase {
+
+    /**
+     * Processes the course and returns a response.
+     * The response tells either
+     *  - that the subplugin is finished processing.
+     *  - that the subplugin is not yet finished processing.
+     *  - that a rollback for this course is necessary.
+     * @param int $processid of the respective process.
+     * @param int $instanceid of the step instance.
+     * @param mixed $course to be processed.
+     * @return step_response
+     * @throws \coding_exception
+     * @throws \dml_exception
+     */
+    public function process_course($processid, $instanceid, $course) {
+        // TODO Write db function insert_db($type).
+        global $DB;
+        $context = context_course::instance($course->id);
+        $userrecords = get_users_by_capability($context, 'lifecyclestep/notifystudents:choice');
+        foreach ($userrecords as $userrecord) {
+            $record = new \stdClass();
+            $record->touser = $userrecord->id;
+            $record->courseid = $course->id;
+            $record->instanceid = $instanceid;
+            $record->emailtype = 0;
+            $DB->insert_record('lifecyclestep_notifystudents', $record);
+        }
+        $userrecords = get_enrolled_users($context, '', 0, '*');
+        foreach ($userrecords as $userrecord) {
+            if (user_has_role_assignment($userrecord->id, 5)) {
```

Comment:
I think this would also need $context as a third parameter, however this doesn't seem like the best way to get students.

 @NinaHerrmann @Laur0r, how would you do that? Add an additional capability for students and do it the same way like in line 64?

Reply:
My reasoning why I don't really like the hardcoded roleid: I think moodle roles are customizable, and I wouldn't expect that the emails definitely have to be send to role 5 :)

Reply:
![image](https://user-images.githubusercontent.com/74201118/136179598-af20b363-7a44-4a7f-9b94-30241f5826ea.png)

Is it that simple or do I have to find a capability I can clone? I couldn't find any capabilities in the moodle repo that would match these requirements.

Reply:
"students" is supposed to be "student" which was copied from the email step capability.